modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/Services/ServicesClient.cs

```diff
@@ -202,7 +216,123 @@ public IServicesClient WithServerKeepBinary()
                         : ctx.Reader;
 
                     return reader.ReadObject<object>();
-                });
+                }, 
+                GetServiceTopology(serviceName));
+        }
+
+        /// <summary>
+        /// Provides actual known service topology or empty list if: service topology is not enabled, not supported or
+        /// not received yet.
+        /// </summary>
+        private IList<Guid> GetServiceTopology(string serviceName)
+        {
+            if (_servicesTopologies == null || !_ignite.Socket.GetSocket().Features.HasFeature(ClientBitmaskFeature.ServiceTopology))
+                return null;
+            
+            return _servicesTopologies.GetOrAdd(serviceName, s => new ServiceTopology(serviceName, this)).GetAndUpdate();
+        }
+        
+        /// <summary>
+        /// Keeps topology of certain service and its update progress.
+        /// </summary>
+        private class ServiceTopology
+        {
+            /** Service name. */
+            private readonly string _svcName;
+            
+            /** Ignite services. */
+            private readonly ServicesClient _svcClient;
+
+            /** Flag of topology update progress. */
+            private int _updateInProgress;
+
+            /** Time of the last received topology. */
+            private long _lastUpdateRequestTime;
+            
+            /** UUIDs of the nodes with at least one service instance. */
+            private volatile IList<Guid> _nodes;
+
+            /** Last cluster topology version when current service topology was actual. */
+            private long _lastAffTop;
+
+            /** */
+            internal ServiceTopology(string name, ServicesClient svcClient)
+            {
+                _svcName = name;
+                _svcClient = svcClient;
+            }
+
+            /// <summary>
+            /// Asynchronously updates the service topology.
+            /// </summary>
+            private async void UpdateTopologyAsync()
```

Comment:
```suggestion
            private async Task UpdateTopologyAsync()
```

`async void` is not recommended:
* https://learn.microsoft.com/en-us/archive/msdn-magazine/2013/march/async-await-best-practices-in-asynchronous-programming
* https://learn.microsoft.com/en-us/archive/msdn-magazine/2013/march/async-await-best-practices-in-asynchronous-programming